// razy/dist/server/server.js
/**
 * @fileOverview 服务器入口
 * @author Max
 **/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var express = require("express");
require("./config");
var cookieParser = require("cookie-parser");
var compression = require("compression");
var session = require("express-session");
var bodyParser = require("body-parser");
var jr = require("./utils/json-result");
var page_1 = require("./page");
var env = require("./utils/env-detect");
var promise_extension_1 = require("../utils/promise-extension");
promise_extension_1.extend();
var less_loader_1 = require("./utils/less-loader");
global._importLess = less_loader_1.importFile;
var iso_request_1 = require("../utils/iso-request");
var retina_1 = require("../utils/retina");
retina_1.default(__RETINA_DEFAULT_RATIO__);
global.ret = retina_1.ret;
var app = express();
function start(params, interceptor) {
    var requestMethods = iso_request_1.default(params.dataFlagResolver);
    global._http = requestMethods.http;
    global._https = requestMethods.https;
    jr.init({
        error: params.resultObjErrorConstructor,
        success: params.resultObjSuccessConstructor
    });
    page_1.init(params.reducerRoot, params.routes, params.createStore);
    app.use(compression());
    app.use(cookieParser());
    app.use(session({
        secret: 'keyboard cat',
        resave: true,
        saveUninitialized: true
    }));
    app.use(bodyParser.urlencoded({ extended: false }));
    app.use(bodyParser.json());
    app.all('*', function (req, res, next) {
        res.setHeader("Access-Control-Allow-Origin", __CROSSDOMAIN__);
        next();
    });
    // serve static files
    app.use("/" + __DIST_PATH__, express.static(__DIST_PATH__));
    app.use("/" + __DIST_PATH__, function (req, res, next) {
        res.status(404).end('Static File Not Found');
    });
    // stub
    if (env.dev()) {
        var StubMiddleware = require('./stub');
        app.use('/stub', StubMiddleware.router);
    }
    // interceptor
    if (interceptor)
        interceptor(app);
    // page
    app.use('/', page_1.default);
    app.listen(__PORT__);
    console.log("listen to " + __PORT__);
}
exports.start = start;
