// razy/dist/utils/iso-request.js
/**
 * @fileOverview isomorphic request method, compatible both in server and client side
 * @author Max
 */
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function getRequestMethod(dataFlagResolver) {
    if (typeof window === 'undefined') {
        var _a = require('./http'), initHTTP = _a.initHTTP, initHTTPS = _a.initHTTPS;
        return {
            http: initHTTP({
                hostname: __API_SERVER_HTTP_HOSTNAME__,
                port: __API_SERVER_HTTP_PORT__,
                stubHostname: __STUB_SERVER_HTTP_HOSTNAME__,
                stubPort: __STUB_SERVER_HTTP_PORT__
            }, dataFlagResolver),
            https: initHTTPS({
                hostname: __API_SERVER_HTTP_HOSTNAME__,
                port: __API_SERVER_HTTP_PORT__,
                stubHostname: __STUB_SERVER_HTTP_HOSTNAME__,
                stubPort: __STUB_SERVER_HTTP_PORT__
            }, dataFlagResolver)
        };
    }
    else {
        var initFetch = require('./fetch').initFetch;
        var fetch_1 = initFetch({
            hostname: __API_SERVER_FETCH_HOSTNAME__,
            port: __API_SERVER_FETCH_PORT__,
            stubHostname: __STUB_SERVER_FETCH_HOSTNAME__,
            stubPort: __STUB_SERVER_FETCH_PORT__
        }, dataFlagResolver);
        return {
            http: fetch_1,
            https: fetch_1
        };
    }
}
exports.default = getRequestMethod;
