// razy/dist/utils/validator.js
/**
 * @fileOverview validator utils
 * @author Max
 **/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var isMobilePhone = require("validator/lib/isMobilePhone");
var isInt = require("validator/lib/isInt");
var isFloat = require("validator/lib/isFloat");
var isURL = require("validator/lib/isURL");
exports.URLValidator = isURL;
/**
 * only for basic types
 */
exports.notEmptyValidator = function (value) { return value !== '' &&
    value !== undefined &&
    value !== null &&
    !(isNaN(value) && typeof value === 'number') &&
    !(Array.prototype.isPrototypeOf(value) && value.length === 0) &&
    !(typeof value === 'object' && isEmptyObj(value)); };
exports.phoneValidator = function (value) { return isMobilePhone(value + '', 'zh-CN'); };
exports.integerValidator = function (value) { return typeof value === 'number' && isInt(value + ''); };
exports.floatValidator = function (value) { return typeof value === 'number' && isFloat(value + ''); };
exports.booleanValidator = function (value) { return typeof value === 'boolean'; };
exports.arrayValidator = function (value) { return Array.prototype.isPrototypeOf(value); };
function isEmptyObj(obj) {
    for (var i in obj) {
        return false;
    }
    return true;
}
