// razy/dist/client/components/base-page.js
/**
 * @fileOverview base class for page component
 * @author Max
 */
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var html_manager_1 = require("../../server/bootstrap/html-manager");
var htmlManager = new html_manager_1.default();
var BaseComponent = (function (_super) {
    __extends(BaseComponent, _super);
    function BaseComponent(props) {
        return _super.call(this, props) || this;
    }
    BaseComponent.prototype.getInitDataAction = function (props, force) {
        if (force === void 0) { force = false; }
        if (force || !props.dataInited) {
            var actions = this.getInitDataActionImp(props);
            if (isAnyArray(actions))
                return actions;
        }
        return null;
    };
    BaseComponent.prototype.componentDidMount = function () {
        this.setUpPage(htmlManager);
        var dispatch = this.props.dispatch;
        var actions = this.getInitDataAction(this.props);
        if (actions)
            actions.forEach(function (item) { return dispatch(item); });
        this.setState({
            client: true,
            bodyHeight: window.innerHeight
        });
    };
    return BaseComponent;
}(React.Component));
function isAnyArray(array) {
    return array !== undefined;
}
exports.default = BaseComponent;
